Start transliterating ASL semantics rules

jalglave opened the pull request:
hi @HadrienRenaud please could you try to generate a webpage from Interpreter.mli and add the recipe to do so to this PR? Then we can merge, as soon as possible.
Thanks in advance,
Jade

jalglave commented:
hi @HadrienRenaud and @maranget please could we merge this?
Thanks,
Jade

HadrienRenaud commented on asllib/backend.mli:
```diff
@@ -64,15 +65,17 @@ module type S = sig
+  val bind_par : 'a m -> 'b m -> ('a * 'b) m
+  (** Monadic product operation, two monads are combined "in parallel".*)
+
+  val appl: 'a m -> ('a -> 'b) -> 'b m
```
I think it should be called `appl_data`, and that we should add a documentation comment:
```suggestion
  val appl_data: 'a m -> ('a -> 'b) -> 'b m
  (** Applicative map. 
  
      Creates a data dependency between the output events and the input events of the argument in the resulting monad. *)
```

HadrienRenaud commented on herd/ASLSem.ml:
```diff
@@ -507,7 +509,8 @@ module Make (C : Config) = struct
         let bind_ctrl = M.bind_ctrl_seq_data
-        let prod = M.( >>| )
+        let bind_par = M.( >>| )
+        let appl m f = m >>= fun v -> return (f v)
```
@maranget this is out of scope for this PR but maybe it would be interesting to implement this into `herd/monad.mli`?

HadrienRenaud commented on asllib/instrumentation.ml:
```diff
@@ -79,16 +87,15 @@ module Rule = struct
-      GetBitFields;
-      GetBitField;
+      EGetBitField;
+      EGetBitFields;
```
I think that you should also add the instrumentation rules that you've added to this list:
```suggestion
      EGetBitFields;
      EUnknown;
      EPattern;
      EGetArray;
      ECondSimple;
      EUndefIdent;
```

HadrienRenaud commented on asllib/Interpreter.ml:
```diff
@@ -65,35 +67,113 @@ module Make (B : Backend.S) (C : Config) = struct
-  let ( let*| ) = B.bind_seq
+  (* Parallel bind *)
+  let ( and* ) = B.bind_par
+  (* could and* be called par? *)
```
We can rename `bind_par` to be called `par`, but to use the syntactic feature of the `and*`, we can only name it something that is formed by "and" followed by some operators (see [the manual](https://v2.ocaml.org/manual/bindingops.html#and-operator))
If we dropped the syntactic sugar of the `and*`, then why not name it `bind_par` (or its name in `Backend`) not to confuse people?

HadrienRenaud commented on asllib/Interpreter.ml:
```diff
@@ -304,53 +331,70 @@ module Make (B : Backend.S) (C : Config) = struct
-    | E_Cond (econd, e2, e3) ->
+        return_normal (v, env)
+
+    | E_Cond (econd, o1, o2) ->
```
Why `o` and not `e`?

HadrienRenaud commented on asllib/Interpreter.ml:
```diff
@@ -304,53 +331,70 @@ module Make (B : Backend.S) (C : Config) = struct
-            normal (v, env) |: Rule.ELocalVar
-        | NotFound -> fatal_from e @@ Error.UndefinedIdentifier x)
+            return_normal (v, env) |: Rule.EGlobalVar
+        | NotFound -> fatal_from e @@ Error.UndefinedIdentifier x) |: Rule.EUndefIdent 
```
```suggestion

        | NotFound -> fatal_from e @@ Error.UndefinedIdentifier x |: Rule.EUndefIdent)
```

HadrienRenaud commented on asllib/Interpreter.ml:
```diff
@@ -369,33 +414,45 @@ module Make (B : Backend.S) (C : Config) = struct
         | Some i ->
             let* v = B.get_index i v_array in
-            normal (v, env))
+            return_normal (v, env)) |: Rule.EGetArray 
```
```suggestion
            return_normal (v, env) |: Rule.EGetArray)
```

HadrienRenaud commented on asllib/Interpreter.ml:
```diff
@@ -369,33 +414,45 @@ module Make (B : Backend.S) (C : Config) = struct
         | None -> fatal_from e (Error.UnsupportedExpr e_index)
```
```suggestion
        | None ->
          (* TODO: create a proper runtime error for this.
             It should be caught at type-checking, but still. *) 
          fatal_from e (Error.UnsupportedExpr e_index)
```

HadrienRenaud reviewed (commented):
Happy for this to go in 🙂 

I don't really understand why we need both the documentation in `Interpreter.mli` and in `ASLSemanticsReference.mld` but it's fine for me now.

jalglave commented:
Thanks @HadrienRenaud. I simply had forgotten to add the new Interpreter.mli (sans documentation comments); it's rectified now. I think I have addressed all your comments in my latest commit - please could you check?
Thanks again,
Jade

HadrienRenaud commented:
All right for me now!

maranget commented:
Hi would you mind if I squash all commits into one ? With message: "Documenting ASL semantics, a start"

jalglave commented:
Please can you make the message "Start transliterating ASL semantics rules"?
Otherwise, fine by me to squash it all into one.
Thanks,
Jade
________________________________
From: Luc Maranget ***@***.***>
Sent: 08 September 2023 15:59
To: herd/herdtools7 ***@***.***>
Cc: jalglave ***@***.***>; Author ***@***.***>
Subject: Re: [herd/herdtools7] Start transliterating ASL semantics rules (PR #644)


⚠ Caution: External sender


Hi would you mind if I squash all commits into one ? With message: "Documenting ASL semantics, a start"

—
Reply to this email directly, view it on GitHub<https://github.com/herd/herdtools7/pull/644#issuecomment-1711809410>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABOVDYQDP3XRF3YFTM7VJEDXZMXG5ANCNFSM6AAAAAA4NOXJJM>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

maranget commented:
Fine, let's go.
